pkg/server/cstorvolumeconfig/backup_endpoint.go

```diff
@@ -0,0 +1,545 @@
+	// If given backup is the last backup of scheduled backup (lastbkp.Spec.PrevSnapName == backup) or
+	// completedBackup doesn't have successful backup(len(lastbkp.Spec.PrevSnapName) == 0) then we will delete the lastbkp CR
+	// Deleting this CR make sure that next backup of the schedule will be full backup
+	if lastbkp != nil && (lastbkp.Spec.PrevSnapName == backup || len(lastbkp.Spec.PrevSnapName) == 0) {
+		err := bOps.clientset.OpenebsV1alpha1().CStorCompletedBackups(ns).Delete(lastCompletedBackup, &metav1.DeleteOptions{})
+		if err != nil && !k8serror.IsNotFound(err) {
+			return errors.Wrapf(err, "failed to delete last-completed-backup=%s resource", lastCompletedBackup)
+		}
+	}
+
+	snapshot := snapshot.Snapshot{
+		VolumeName:   volname,
+		SnapshotName: backup,
+		Namespace:    ns,
+		SnapClient:   bOps.snapshotter,
+	}
+	// Snapshot Name and backup name are same
+	_, err = snapshot.DeleteSnapshot(bOps.clientset)
+	if err != nil {
+		return errors.Wrapf(err, "snapshot deletion failed")
+	}
+
+	cstorBackup := backup + "-" + volname
+	err = bOps.clientset.OpenebsV1alpha1().CStorBackups(ns).Delete(cstorBackup, &metav1.DeleteOptions{})
+	if err != nil && !k8serror.IsNotFound(err) {
+		return errors.Wrapf(err, "failed to delete cstorbackup: %s resource", cstorBackup)
+	}
+	return nil
+}
+
+// backupCreateRequestValidations validates the backup create request
+func backupCreateRequestValidations(backUp *openebsapis.CStorBackup) error {
+	// backup name is expected
+	if len(strings.TrimSpace(backUp.Spec.BackupName)) == 0 {
+		return CodedError(400, fmt.Sprintf("Failed to create backup: missing backup name "))
+	}
+
+	// namespace is expected
+	if len(strings.TrimSpace(backUp.Namespace)) == 0 {
+		return CodedError(400, fmt.Sprintf("Failed to create backup '%v': missing namespace", backUp.Spec.BackupName))
+	}
+
+	// volume name is expected
+	if len(strings.TrimSpace(backUp.Spec.VolumeName)) == 0 {
+		return CodedError(400, fmt.Sprintf("Failed to create backup '%v': missing volume name", backUp.Spec.BackupName))
+	}
+
+	// backupIP is expected for remote snapshot
+	if !backUp.Spec.LocalSnap && len(strings.TrimSpace(backUp.Spec.BackupDest)) == 0 {
+		return CodedError(400, fmt.Sprintf("Failed to create backup '%v': missing backupIP", backUp.Spec.BackupName))
+	}
+
+	// snapshot name is expected
+	if len(strings.TrimSpace(backUp.Spec.SnapName)) == 0 {
+		return CodedError(400, fmt.Sprintf("Failed to create backup '%v': missing snapName", backUp.Spec.BackupName))
+	}
+	return nil
+}
+
+// getVolumeIP fetches the cstor target service IP Address
+func getVolumeIP(volumeName string, clientset clientset.Interface) (string, error) {
+	namespace := getOpenEBSNamespace()
+
+	// Fetch the corresponding cstorvolume
+	cstorvolume, err := clientset.CstorV1().CStorVolumes(namespace).
+		Get(volumeName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	return cstorvolume.Spec.TargetIP, nil
```

Comment:
Is there not a possibility of` cstorvolume.Spec.TargetIP` be empty?

Reply:
No

Reply:
If it is empty snap creation will fail. So backup will fail